### PR TITLE
Cobra CLI

### PR DESCRIPTION
**Pull request by happyRip**
Implemented Cobra CLI based on response to my previous [pull request](https://github.com/maaslalani/typer/pull/6)

Usage:

* `typer` is used to generate random words by default.

* `--file` / `-f` followed by filename is used to read input from file instead.

* `--length` / `-l` followed by an integer is used to set number of max words, by default is set to max int value.

* `--capital` / `-c` is used to indicate that the user wants to include capital letters from text, by default capital letters are changed to lowercase.

* `--punctuation` / `-p` is used to indicate that the user wants to include punctuation from text, by default punctuation is removed.

All groups of whites pace characters are replaced with a single space character by design.

The app lacks description to be included with Cobra, it needs your touch imho.

Licensing was changed to Cobra's license, don't know how to resolve that tbf.

edit: Wasn't able to link related [issue](https://github.com/maaslalani/typer/issues/7)

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	if cfgFile != "" {
+		// Use config file from the flag.
+		viper.SetConfigFile(cfgFile)
+	} else {
+		// Find home directory.
+		home, err := homedir.Dir()
+		cobra.CheckErr(err)
+
+		// Search config in home directory with name ".typer" (without extension).
+		viper.AddConfigPath(home)
```
Comment (maaslalani) [inline cmd/root.go:121]: Are we also able to search for ~/.config/typer

File: pkg/utility/utility.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/tjarratt/babble"
+)
+
+const (
+	MaxUint = ^uint(0)
+	MaxInt  = int(MaxUint >> 1)
```
Comment (maaslalani) [inline pkg/utility/utility.go]: This looks like it's used for the default value of the length flag, can we just define a defaultLength variable and use that instead for the default value?

File: pkg/typer/typer.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	}
+
+	if !f.Punctuation {
+		s, err = u.RemoveNonAlpha(s)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	if f.Length <= 0 {
```
Comment (maaslalani) [inline pkg/typer/typer.go:91]: Should we also check for > maxLength?

File: LICENSE [hunk 4]
```diff
@@ -1,21 +1,202 @@
...
-
-THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-SOFTWARE.
+
+                                 Apache License
```
Comment (maaslalani) [inline LICENSE]: I'm curious why the license changed, did you use the cobra generator? I would like to keep MIT if possible.

Comment (happyRip) [reply LICENSE]: Apache license states: 

>  (a) You must give any other recipients of the Work or
          Derivative Works a copy of this License; 

so I kept this license. TBF don't know how to resolve it.

File: cmd/root.go (hunk 1 above)
Comment (happyRip) [reply cmd/root.go:121]: I'm not yet familiar with `Viper`, can look into that later.

File: LICENSE (hunk 4 above)
Comment (maaslalani) [reply LICENSE]: Hmmm, I really doubt that using cobra (as a distributed package) requires you to change your license, (forking and changing it definitely would). For example, [GitHub's cli uses cobra](https://github.com/cli/cli/blob/5a2ec54685806a6576bdc185751afc09aba44408/cmd/gh/main.go#L31) and [is licensed under MIT](https://github.com/cli/cli/blob/trunk/LICENSE)

**Comment by happyRip** [comment]
I addressed most of the review with the changes.
The `Viper` is out of my reach for now.

**Comment by maaslalani** [comment]
`go run main.go` didn't work?

image

**Comment by maaslalani** [comment]
`go run main.go -l 100` worked for me 👍 

But I don't think the default length should be `MaxInt`, maybe change it to 20?

**Comment by maaslalani** [comment]
Also, we will need to be able to `ctrl+c` to exit the typing test

**Comment by happyRip** [comment]
I've adjusted word limits. `DefaultLength` was set to `20` words and `MaxLength` was set to `500` words, which seems reasonable, but can be adjusted further.

**Comment by maaslalani** [comment]
Changes merged in https://github.com/maaslalani/typer/pull/9

Thanks so much for contributing @happyRip ❤️  I really appreciate it!

**Pull request closed without merging**